### PR TITLE
Add a collection of new unit tests for the range tree.

### PR DESCRIPTION
**Pull request by BramGruneir**
Next up is adding deletions/merges

### CODE REVIEW

File: storage/client_range_tree_test.go [hunk 1]
```diff
@@ -584,10 +592,125 @@ func TestInsertLeft(t *testing.T) {
...
 		t.Fatal(err)
 	}
 }
+
+// Compare implements the llrbComparable interface for keys.
+func (k Key) Compare(b llrb.Comparable) int {
+	return bytes.Compare(k, b.(Key))
+}
+
+// nodeSame compares a boigo node and a range tree node to determine if both
```
Comment (tschottdorf) [inline storage/client_range_tree_test.go]: s/boigo/biogo/

Comment (BramGruneir) [inline storage/client_range_tree_test.go]: fixed (in two places)

File: storage/client_range_tree_test.go [hunk 2]
```diff
@@ -584,10 +592,125 @@ func TestInsertLeft(t *testing.T) {
...
+
+// Compare implements the llrbComparable interface for keys.
+func (k Key) Compare(b llrb.Comparable) int {
+	return bytes.Compare(k, b.(Key))
+}
+
+// nodeSame compares a biogo node and a range tree node to determine if both
+// contain the same values in the same order.  It recursively calls itself on
+// both children if they exist.
+func nodeSame(db *client.KV, biogoNode *llrb.Node, key *proto.Key) error {
```
Comment (spencerkimball) [inline storage/client_range_tree_test.go]: These methods appear to be very similar to the ones for comparing to expected trees at the top of this test file. Why don't you rename this compareBiogoNode for clarity?

File: storage/range_tree_test.go [hunk 3]
```diff
@@ -46,3 +68,443 @@ func TestIsRed(t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+		if !reflect.DeepEqual(actualLeft, test.expectedLeft) {
+			t.Errorf("%d: Expected Left Node is not the same as the actual.\nExpected: %+v\nActual: %+v\n", i, test.expectedLeft, actualLeft)
+		}
+	}
+}
+
+// TestRotateLeft ensures that left rotations occur corectly.
```
Comment (spencerkimball) [inline storage/range_tree_test.go]: s/corectly/correctly/

File: storage/range_tree_test.go [hunk 4]
```diff
@@ -18,12 +18,34 @@
...
+		RootKey: *rootKey,
+	}
+	tc := &treeContext{
+		db:    nil,
+		tree:  root,
+		dirty: false,
+		nodes: map[string]cachedNode{},
+	}
+	for _, node := range nodes {
+		// We don't use setNode here to ensure updated is false.
```
Comment (spencerkimball) [inline storage/range_tree_test.go]: s/updated/dirty/

File: storage/range_tree_test.go [hunk 5]
```diff
@@ -46,3 +68,443 @@ func TestIsRed(t *testing.T) {
 		}
 	}
 }
+
+// TestFlip ensures that flips function correctly.
+func TestFlip(t *testing.T) {
```
Comment (spencerkimball) [inline storage/range_tree_test.go]: Given all these tests in this file (which are better as unittests than the ones in client_range_tree_test), as well as the randomized test just added, my feeling is we should drop the super verbose tests in client_range_tree_test which verify specific tree structures. I think we've overtested now between the two and those super verbose ones would be my choice to eliminate as they're going to be hard for anyone else to maintain.

**Comment by spencerkimball** [comment]
LGTM

File: storage/client_range_tree_test.go (hunk 2 above)
Comment (BramGruneir) [inline storage/client_range_tree_test.go]: Good Call.  I was struggling with a good name for them.

File: storage/range_tree_test.go (hunk 3 above)
Comment (BramGruneir) [inline storage/range_tree_test.go]: fixed

File: storage/range_tree_test.go (hunk 4 above)
Comment (BramGruneir) [inline storage/range_tree_test.go]: fixed

File: storage/range_tree_test.go (hunk 5 above)
Comment (BramGruneir) [inline storage/range_tree_test.go]: I'd rather leave it in for now.  At least until after I get the deletion and parent key work finished.

**Comment by BramGruneir** [comment]
Thanks Spencer!

**Pull request merged**